services/agents/mongodb.go

```diff
@@ -41,20 +41,62 @@ var (
 	newMongoExporterPMMVersion = version.MustParse("2.9.99")
 	v2_24_99                   = version.MustParse("2.24.99")
 	v2_25_99                   = version.MustParse("2.25.99")
+	v2_26_99                   = version.MustParse("2.26.99")
 )
 
+func getMongoExporterFiles(exporter *models.Agent, pmmAgentVersion *version.Parsed) map[string]string {
+	files := exporter.Files()
+
+	// The mongoDB exporter prior 2.26 use exporter_shared and gets basic auth config from env.
+	// Starting with pmm 2.27, the exporter uses Prometheus Web Toolkit and needs a config file
+	// with the basic auth users.
+	if pmmAgentVersion.Less(v2_26_99) { // <= 2.27
+		delete(files, "webConfigPlaceholder")
+	}
+	if len(files) == 0 {
+		return nil
+	}
+
+	return files
+}
+
+func getMongoExporterEnv(service *models.Service, exporter *models.Agent, pmmAgentVersion *version.Parsed, database string) []string {
+	tdp := exporter.TemplateDelimiters(service)
+
+	env := []string{
+		fmt.Sprintf("MONGODB_URI=%s", exporter.DSN(service, time.Second, database, tdp)),
+	}
+
+	if pmmAgentVersion.Less(v2_26_99) { // <= 2.27
+		env = append(env, fmt.Sprintf("HTTP_AUTH=pmm:%s", exporter.GetAgentPassword()))
+	}
+
+	return env
+}
+
 // mongodbExporterConfig returns desired configuration of mongodb_exporter process.
 func mongodbExporterConfig(service *models.Service, exporter *models.Agent, redactMode redactMode,
 	pmmAgentVersion *version.Parsed,
 ) *agentpb.SetStateRequest_AgentProcess {
 	tdp := exporter.TemplateDelimiters(service)
 
+	database := ""
+	if exporter.MongoDBOptions != nil {
+		database = exporter.MongoDBOptions.AuthenticationDatabase
+	}
+
+	env := getMongoExporterEnv(service, exporter, pmmAgentVersion, database)
+	files := getMongoExporterFiles(exporter, pmmAgentVersion)
+
 	var args []string
 	// Starting with PMM 2.10.0, we are shipping the new mongodb_exporter
 	// Starting with PMM 2.25.0, we changes the discovering-mode making it to discover all databases.
 	// Until now, discovering mode was not workign properly and was enabled only if mongodb.collstats-colls=
 	// was specified in the command line.
 	switch {
+	case !pmmAgentVersion.Less(v2_26_99): // >= 2.27
```

Comment:
new code looks hard to read and understand, we had only one place where we had all version comparison and changes related to them and now we have 4 places where we change values based on different variables.

Reply:
the same comments for node_exporter